### PR TITLE
Fix OCB partial block issue

### LINKED ISSUE #2273: aes ocb bug
This is for OpenSSL 1.1.0c

Encrypting or decrypting using AES OCB is broken if the input data size is not a multiple of the block size (16 bytes).  The problem is in crypto/evp/e_aes.c in the function aes_ocb_cipher in this block:
```
2560         if (*buf_len) {
2561             unsigned int remaining;
2562
2563             remaining = 16 - (*buf_len);
2564             if (remaining > len) {
2565                 memcpy(buf + (*buf_len), in, len);
2566                 *(buf_len) += len;
2567                 return 0;
2568             }
2569             memcpy(buf + (*buf_len), in, remaining);
2570
2571             /*
2572              * If we get here we've filled the buffer, so process it
2573              */
2574             len -= remaining;
2575             in += remaining;
2576             if (out == NULL) {
2577                 if (!CRYPTO_ocb128_aad(&octx->ocb, buf, 16))
2578                     return -1;
2579             } else if (EVP_CIPHER_CTX_encrypting(ctx)) {
2580                 if (!CRYPTO_ocb128_encrypt(&octx->ocb, buf, out, 16))
2581                     return -1;
2582             } else {
2583                 if (!CRYPTO_ocb128_decrypt(&octx->ocb, buf, out, 16))
2584                     return -1;
2585             }
2586             written_len = 16;
2587             *buf_len = 0;
2588             out += 16; /*  <=== Missing line of code */
2589         }
```

Here we are building a 16 byte block by combining previously supplied data which was less than 16 bytes with new input data.  The problem is that we do not advance the output pointer, which leads to the data being overwritten.

Adding line 2588 fixes the bug.

Seen when encrypting data with the EVP_ase_256_ocb cipher when one or more input data block sizes are not a multiple of 16 and at least one said data block is NOT the last block.

Norm Green

### PR DESCRIPTION
**Pull request by mattcaswell**
##### Checklist

- [ ] documentation is added or updated
- [x] tests are added or updated
- [x] CLA is signed

##### Description of change

This fixes a bug in OCB where if an "Update" function is called with a partial block of data, then a subsequent "Update" will use the partial data already supplied and write out the block as expected, but then overwrite it with the next full block of output.

I also updated the tests to catch this case.

Fixes #2273

### CODE REVIEW

**Comment by mattcaswell** [comment]
Writing the test for this threw up a few things. Firstly XTS and wrap do not seem to be able to cope with unusual (partial block) input lengths. This is possibly expected as these modes are "special".

I also found that CCM also could not handle unusual lengths. I am less sure whether this is expected behaviour or not.

Perhaps @snhenson can shed some light on the above?

I also discovered that the `is_partially_overlapping()` check fails if you supply partial blocks of data to an update function whilst using in place encryption. This looks like a possible bug to me. @dot-asm your view?

**Comment by dot-asm** [comment]
Problem with suggested evp_test.c modification is that there is a number of vectors in evptests.txt with lengths specifically chosen to exercise corner cases [customarily] in vector code paths. And if you chip off bytes for EVP to handle, some of intended code paths won't be exercised. Well, as of now they will be exercised in in-place case, but one can argue that other cases are relevant too. What I'm trying to say is that fragmented input should probably be *additional* separate test, not replacement...

File: crypto/evp/e_aes.c [hunk 1]
```diff
@@ -2574,21 +2574,25 @@ static int aes_ocb_cipher(EVP_CIPHER_CTX *ctx, unsigned char *out,
...
-                if (!CRYPTO_ocb128_decrypt(&octx->ocb, buf, out, 16))
+                if (!CRYPTO_ocb128_decrypt(&octx->ocb, buf, out,
+                                           AES_BLOCK_SIZE))
                     return -1;
             }
-            written_len = 16;
+            written_len = AES_BLOCK_SIZE;
             *buf_len = 0;
+            if (out != NULL)
+                out+= AES_BLOCK_SIZE;
```
Comment (richsalz) [inline crypto/evp/e_aes.c]: space before the operator.  yes, a real obsessive-compulsive behavior kind of nit, but this is about OCB innit?

File: test/evp_test.c [hunk 2]
```diff
@@ -983,15 +983,67 @@ static int cipher_test_enc(struct evp_test *t, int enc,
...
+                                  cdat->aad_len))
+                goto err;
+        } else {
+            /*
+             * Supply the AAD in chunks less than the block size where possible
+             */
+            if (cdat->aad_len > 0) {
+                if (!EVP_CipherUpdate(ctx, NULL, &chunklen, cdat->aad, 1))
+                    goto err;
+                donelen += 1;
```
Comment (richsalz) [inline test/evp_test.c]: i get the += for parallelism, but it does jangle to not see ++

File: crypto/evp/e_aes.c [hunk 3]
```diff
@@ -2557,10 +2557,10 @@ static int aes_ocb_cipher(EVP_CIPHER_CTX *ctx, unsigned char *out,
          * If we've got a partially filled buffer from a previous call then
          * use that data first
          */
-        if (*buf_len) {
+        if (*buf_len > 0) {
             unsigned int remaining;
 
-            remaining = 16 - (*buf_len);
+            remaining = AES_BLOCK_SIZE - (*buf_len);
```
Comment (richsalz) [inline crypto/evp/e_aes.c]: yeah for using defines!

**Comment by kroeckx** [comment]
It looks broken for in_len == 3 and aad_len == 3

File: crypto/evp/e_aes.c (hunk 1 above)
Comment (mattcaswell) [inline crypto/evp/e_aes.c]: Fixed.

File: test/evp_test.c (hunk 2 above)
Comment (mattcaswell) [inline test/evp_test.c]: Changed.

**Comment by mattcaswell** [comment]
> It looks broken for in_len == 3 and aad_len == 3

How? I don't see it??

**Comment by mattcaswell** [comment]
> Problem with suggested evp_test.c modification is that there is a number of vectors in evptests.txt with lengths specifically chosen to exercise corner cases [customarily] in vector code paths. And if you chip off bytes for EVP to handle, some of intended code paths won't be exercised. Well, as of now they will be exercised in in-place case, but one can argue that other cases are relevant too. What I'm trying to say is that fragmented input should probably be additional separate test, not replacement...

I changed things around a little so that it is now an additional test rather than a replacement.

**Comment by mattcaswell** [comment]
New commits pushed addressing all issues.

I also added a new commit to deal with the issue of `is_partially_overlapping()` failing when using fragmented input and in place encryption. Perhaps @dot-asm might want to take a look.

**Comment by kroeckx** [comment]
On Tue, Jan 24, 2017 at 05:04:08AM -0800, mattcaswell wrote:
> > It looks broken for in_len == 3 and aad_len == 3
> 
> How? I don't see it??

I got confused. In any case, it's hard to follow code. Please
consider adjusting in and in_len instead of using donelen.

**Comment by mattcaswell** [comment]
> I got confused. In any case, it's hard to follow code. Please
consider adjusting in and in_len instead of using donelen.

I added an extra commit to do that - although it only makes sense in the non-aad portion of the code (the aad portion uses non-local variables instead of `in` and `in_len` so it doesn't make sense to change them).

**Comment by mattcaswell** [comment]
I'd really like to get this in the release tomorrow if someone is able to review.

File: crypto/evp/evp_enc.c [hunk 4]
```diff
@@ -276,8 +276,7 @@ int EVP_DecryptInit_ex(EVP_CIPHER_CTX *ctx, const EVP_CIPHER *cipher,
 # define PTRDIFF_T size_t
 #endif
 
-static int is_partially_overlapping(const void *ptr1, const void *ptr2,
```
Comment (richsalz) [inline crypto/evp/evp_enc.c]: this could be called ```overlaps```.

**Comment by richsalz** [comment]
I think the assert() in is_partially_overlapping is now completely wrong. Look at the places it's used.

**Comment by richsalz** [comment]
I had already plus-one'd this.  Fix the assert issue and I reconfirm.

**Comment by mattcaswell** [comment]
assert removed. I will merge shortly. And yes this is for master and 1.1.0 as per your earlier comment.

**Comment by mattcaswell** [comment]
Pushed. Thanks.
